File: symmetric-core/src/main/java/org/jumpmind/symmetric/transport/http/HttpIncomingTransport.java
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.jumpmind.symmetric.transport.http;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;

import org.jumpmind.exception.HttpException;
import org.jumpmind.symmetric.common.ParameterConstants;
import org.jumpmind.symmetric.service.IParameterService;
import org.jumpmind.symmetric.service.RegistrationNotOpenException;
import org.jumpmind.symmetric.service.RegistrationPendingException;
import org.jumpmind.symmetric.service.RegistrationRequiredException;
import org.jumpmind.symmetric.transport.AuthenticationException;
import org.jumpmind.symmetric.transport.ConnectionRejectedException;
import org.jumpmind.symmetric.transport.IIncomingTransport;
import org.jumpmind.symmetric.transport.NoContentException;
import org.jumpmind.symmetric.transport.ServiceUnavailableException;
import org.jumpmind.symmetric.transport.SyncDisabledException;
import org.jumpmind.symmetric.transport.TransportUtils;
import org.jumpmind.symmetric.web.WebConstants;

public class HttpIncomingTransport implements IIncomingTransport {
    private HttpTransportManager httpTransportManager;
    private HttpConnection connection;
    private BufferedReader reader;
    private InputStream is;
    private IParameterService parameterService;
    private int httpTimeout;
    private String redirectionUrl;
    private String nodeId;
    private String securityToken;
    private Map<String, String> requestProperties;

    public HttpIncomingTransport(HttpTransportManager httpTransportManager, HttpConnection connection, IParameterService parameterService) {
        this.httpTransportManager = httpTransportManager;
        this.connection = connection;
        this.parameterService = parameterService;
        this.httpTimeout = parameterService.getInt(ParameterConstants.TRANSPORT_HTTP_TIMEOUT);
    }

    public HttpIncomingTransport(HttpTransportManager httpTransportManager, HttpConnection connection, IParameterService parameterService,
            Map<String, String> requestProperties) {
        this(httpTransportManager, connection, parameterService);
        this.requestProperties = requestProperties;
    }

    public HttpIncomingTransport(HttpTransportManager httpTransportManager, HttpConnection connection, IParameterService parameterService,
            String nodeId, String securityToken) {
        this(httpTransportManager, connection, parameterService);
        this.nodeId = nodeId;
        this.securityToken = securityToken;
    }

    @Override
    public String getUrl() {
        return this.connection.getURL().toExternalForm();
    }

    @Override
    public void close() {
        if (connection != null) {
            try {
                connection.close();
            } catch (Exception e) {
            }
        }
        if (reader != null) {
            try {
                reader.close();
            } catch (IOException e) {
            }
            reader = null;
        }
        if (is != null) {
            try {
                is.close();
            } catch (IOException e) {
            }
            is = null;
        }
    }

    @Override
    public boolean isOpen() {
        return reader != null;
    }

    @Override
    public String getRedirectionUrl() {
        return redirectionUrl;
    }

    @Override
    public InputStream openStream() throws IOException {
        applyRequestProperties();
        boolean manualRedirects = parameterService.is(ParameterConstants.TRANSPORT_HTTP_MANUAL_REDIRECTS_ENABLED, true);
        if (manualRedirects) {
            connection = openConnectionCheckRedirects();
        }
        int code = connection.getResponseCode();
        switch (code) {
            case WebConstants.REGISTRATION_NOT_OPEN:
                throw new RegistrationNotOpenException();
            case WebConstants.REGISTRATION_REQUIRED:
                throw new RegistrationRequiredException();
            case WebConstants.REGISTRATION_PENDING:
                throw new RegistrationPendingException();
            case WebConstants.SYNC_DISABLED:
                throw new SyncDisabledException();
            case WebConstants.SC_SERVICE_BUSY:
                throw new ConnectionRejectedException();
            case WebConstants.SC_SERVICE_UNAVAILABLE:
                throw new ServiceUnavailableException();
            case WebConstants.SC_FORBIDDEN:
                httpTransportManager.clearSession(connection);
                throw new AuthenticationException();
            case WebConstants.SC_AUTH_EXPIRED:
                httpTransportManager.clearSession(connection);
                throw new AuthenticationException(true);
            case WebConstants.SC_NO_CONTENT:
                throw new NoContentException();
            case WebConstants.SC_OK:
                httpTransportManager.updateSession(connection);
                is = HttpTransportManager.getInputStreamFrom(connection);
                return is;
            default:
                throw new HttpException(code, "Received an unexpected response code of " + code + " from the server");
        }
    }

    @Override
    public BufferedReader openReader() throws IOException {
        InputStream stream = openStream();
        reader = TransportUtils.toReader(stream);
        return reader;
    }

    @Override
    public Map<String, String> getHeaders() {
        Map<String, String> headers = new LinkedHashMap<String, String>();
        for (String name : connection.getHeaderFields().keySet()) {
            if (name != null) {
                headers.put(name, connection.getHeaderField(name));
            }
        }
        return headers;
    }

    /**
     * This method support redirection from an http connection to an https connection. See
     * {@link http://java.sun.com/j2se/1.4.2/docs/guide/deployment/deployment-guide/upgrade-guide/article-17.html} for more information.
     * 
     * @param connection
     * @return
     * @throws IOException
     */
    private HttpConnection openConnectionCheckRedirects() throws IOException {
        boolean redir;
        int redirects = 0;
        do {
            connection.setInstanceFollowRedirects(false);
            redir = false;
            int stat = connection.getResponseCode();
            if (stat >= 300 && stat <= 307 && stat != 306 && stat != HttpConnection.HTTP_NOT_MODIFIED) {
                URL base = connection.getURL();
                redirectionUrl = connection.getHeaderField("Location");
                URL target = null;
                if (redirectionUrl != null) {
                    target = new URL(base, redirectionUrl);
                }
                connection.disconnect();
                // Redirection should be allowed only for HTTP and HTTPS
                // and should be limited to 5 redirections at most.
                if (target == null || !(target.getProtocol().equals("http")
                        || target.getProtocol().equals("https"))
                        || redirects >= 5) {
                    throw new SecurityException("illegal URL redirect");
                }
                redir = true;
                connection = httpTransportManager.openConnection(target, nodeId, securityToken);
                connection.setConnectTimeout(httpTimeout);
                connection.setReadTimeout(httpTimeout);
                applyRequestProperties();
                redirects++;
            }
        } while (redir);
        return connection;
    }

    protected void applyRequestProperties() throws IOException {
        if (requestProperties != null) {
            connection.setRequestMethod("POST");
            connection.setDoOutput(true);
            StringBuilder sb = new StringBuilder();
            for (Map.Entry<String, String> requestProperty : requestProperties.entrySet()) {
                if (sb.length() != 0) {
                    sb.append("&");
                }
                sb.append(requestProperty.getKey()).append("=");
                if (requestProperty.getValue() != null && !requestProperty.getValue().equals("")) {
                    sb.append(URLEncoder.encode(requestProperty.getValue(), StandardCharsets.UTF_8.name()));
                }
            }
            try (OutputStream os = connection.getOutputStream()) {
                PrintWriter pw = new PrintWriter(new OutputStreamWriter(os, StandardCharsets.UTF_8), true);
                pw.println(sb.toString());
                pw.flush();
            }
        }
    }

    public HttpConnection getConnection() {
        return connection;
    }

    public Map<String, String> getRequestProperties() {
        return requestProperties;
    }

    public void setRequestProperties(Map<String, String> requestProperties) {
        this.requestProperties = requestProperties;
    }
}

File: symmetric-client-clib/src/transport/http/HttpIncomingTransport.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "transport/http/HttpIncomingTransport.h"

static size_t SymHttpIncomingTransport_writeCallback(char *data, size_t size, size_t count, SymDataProcessor *processor) {
    return processor->process(processor, data, size, count);
}

long SymHttpIncomingTransport_process(SymHttpIncomingTransport *this, SymDataProcessor *processor, SymRemoteNodeStatus *status) {
    long httpCode = 0;
    CURLcode rc = CURLE_FAILED_INIT;
    CURL *curl = curl_easy_init();
    if (curl) {
        if (this->parameterService->is(this->parameterService, SYM_PARAMETER_HTTPS_VERIFIED_SERVERS, 1)) {
            curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0);
        }
        if (this->parameterService->is(this->parameterService, SYM_PARAMETER_HTTPS_ALLOW_SELF_SIGNED_CERTS, 1)) {
            curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0);
        }
        curl_easy_setopt(curl, CURLOPT_URL, this->url);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, SymHttpIncomingTransport_writeCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, processor);
        processor->open(processor);
        rc = curl_easy_perform(curl);
        curl_easy_cleanup(curl);
        processor->close(processor);
    } else {
    	SymLog_error("Error cannot initialize curl");
    }
    if (rc == CURLE_OK) {
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &httpCode);
    }

    SymHttpTransportManager_handleCurlRc(rc, httpCode, this->url, status);

    return httpCode;
}

char * SymHttpIncomingTransport_getUrl(SymHttpIncomingTransport *this) {
    return this->url;
}

void SymHttpIncomingTransport_destroy(SymHttpIncomingTransport *this) {
    free(this->url);
    free(this);
}

SymHttpIncomingTransport * SymHttpIncomingTransport_new(SymHttpIncomingTransport *this, char *url, SymParameterService *parameterService) {
    if (this == NULL) {
        this = (SymHttpIncomingTransport *) calloc(1, sizeof(SymHttpIncomingTransport));
    }
    this->parameterService = parameterService;
    SymIncomingTransport *super = &this->super;
    super->getUrl = (void *) &SymHttpIncomingTransport_getUrl;
    super->process = (void *) &SymHttpIncomingTransport_process;
    super->destroy = (void *) &SymHttpIncomingTransport_destroy;

    this->url = url;
    return this;
}
